### DISCUSSION TITLE
Importing tilt brush models on Oculus

### DISCUSSION THREAD

**Discussion by dxdc** [icosa-gallery/open-brush#314] [Q&A] [answered]
Thanks so much for continuing development on this project!

Is there a convenient way to import old TB files on the Oculus (e.g., storing files via specific paths on Sidequest, etc.)? [Icosa Gallery](https://beta.icosa.gallery/) looks very promising, but it seems like it's not integrated just yet. Appreciate any assistance or advice!

**Comment by mikeskydev** [comment, marked as answer]
Hi! Thank you for your support!

Yes, it's super easy to import, the path on Quest is /Open Brush/Sketches, just put your .tilt files in there and OB will recognise them. For reference, Tilt Brush on Quest used to save files at /Tilt Brush/Sketches if you have any to move across from there too. We're considering an option for the user to import from the Tilt Brush folder too.

**Reply by dxdc** [reply]
Thanks so much @merulaxr! I'm going to test this and report back. Is there any estimated timeline for the Icosa Gallery integration?

**Reply by dxdc** [reply]
Also - one other question, if you don't mind - is there a way to import objects (similar to how Poly works) in the meantime, before Icosa Gallery is up? And, what formats are supported? Happy to load all these via Sidequest or other kind of API integration, if supported.

**Reply by mikeskydev** [reply]
We were hoping to have it done already but work got pushed back with our move to OpenXR, sadly. I expect some time at the beginning of the new year now.

> is there a way to import objects

on Quest, you'll need to use our experimental version for now! You can grab the most recent one from the releases page: it'll be in the form OpenBrush_Quest_Experimental_1.0.x.apk, and you'll need to sideload it. We support GLB for Quest so far.

**Reply by dxdc** [reply]
@merulaxr thanks for the assist! I have the latest Experimental APK loaded now.

I added several models into `/Media Library/Models` using a convention of the same folder structure/model name. For example, `Test1/Test1.glb`. After enabling "Advanced" -> "Labs" -> "Local Import", I was able to find these files and import them.

That *does* load the GLB file! However, I have [several examples](https://www.dropbox.com/s/6cps7ss76o9nt9h/sample_glb_tilt.zip?dl=1) where the GLB import does not appear to be working properly. You can view the full files with this online viewer, which shows the correct details: https://gltf-viewer.donmccurdy.com/

Could this be related? https://github.com/icosa-gallery/open-brush/pull/278

I've also included in that batch a tilt file that is not loading properly. In fact, it doesn't even load properly except for the original Oculus that it is loaded on. Is there a cli tool capable of inspecting these tilt files to see the content or assets inside? I'm wondering if there is some linked media that also needs to be transferred. It will only load properly on the original Oculus in Tilt Brush.

Appreciate your help!

**Reply by mikeskydev** [reply]
Yes, the glb importer was part of the original Tilt Brush code and isn't completely to spec! #278 intends to replace it with a much more robust and tested version now that there's some good options in the Unity ecosystem. Apologies that you've having issues with those models.

As for the tilt file, very odd! I'll take a look while loading it in Unity and see if it throws issues. fun fact, a .tilt file is a zip folder, so you can open it with a regular zip tool!

**Reply by dxdc** [reply]
thanks for your help @merulaxr, looking forward to the new GLB importer when that gets merged, will be happy to test it. Also, a plug for having a direct import from sketchfab if doable, that would be very useful!

Regarding the tilt files, I can't figure it out either. I tried another tilt file and it's definitely not a zip file. I wonder if there is some link to Poly hardcoded in the files themselves, and perhaps the data is cached on an existing device? Interestingly, even trying to move them to the `/Tilt Brush/Sketches/` folder on another devices just results in a "Loading Media..." hang screen. [Here's another file](https://www.dropbox.com/s/ihh3yts98981ynv/Untitled_10.tilt?dl=1) if you have any ideas as to what is causing it or where other objects may be linked, welcome any input you'd be able to share!

**Reply by dxdc** [reply]
Btw, I think `/Android/data/com.google.tiltbrush` may be the missing piece here. I'm going to experiment and report back.

**Reply by dxdc** [reply]
> Btw, I think `/Android/data/com.google.tiltbrush` may be the missing piece here. I'm going to experiment and report back.

This was the key to import our existing TiltBrush files into OpenBrush! Inside these folders (both the tiltbrush and openbrush variants) are a `files` folder and `cache` folder. Renaming the `com.google.tiltbrush` folder to the icosa variant allowed all the models and files to be loaded into OpenBrush. Do you know if there's a way to manually load models/assets somehow into this folder structure, and use them directly instead of trying the glb type import? I'm also not sure what file formats may be needed, but happy to continue experimenting.

<img width="796" alt="Screen Shot 2022-11-23 at 9 09 13 PM" src="https://user-images.githubusercontent.com/7200365/203685466-9596c665-8621-42d9-8833-02ff1fc9eea7.png">
